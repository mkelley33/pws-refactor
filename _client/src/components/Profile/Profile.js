import React from 'react';
import { withFormik } from 'formik';
import { toast } from 'react-toastify';
import { Field } from 'formik';
import * as Yup from 'yup';
import compact from 'lodash/compact';

import api from '../../api';
import TextInput from '@components/common/forms/TextInput';
import Checkbox from '@components/common/forms/Checkbox';
import CheckboxGroup from '@components/common/forms/CheckboxGroup';
import styles from './profile.module.css';
import commonStyles from '@components/common/common.module.css';
import ForgotPassword from '../ForgotPassword';
import Modal from '@components/Modal/Modal';
import withAuthentication from '@components/withAuthentication';

const formikEnhancer = withFormik({
  validationSchema: Yup.object().shape({
    firstName: Yup.string().min(2, "C'mon, your first name is longer than that").required('First name is required'),
    lastName: Yup.string().min(2, "C'mon, your last name is longer than that").required('Last name is required'),
    email: Yup.string().email('Invalid email address').required('Email is required'),
    currentPassword: Yup.string(),
    password: Yup.string()
      .min(8, 'Password has to be at least 8 characters!')
      .matches(
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\\$%\\^&\\*])/,
        'Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character'
      ),
    confirmPassword: Yup.string().oneOf([Yup.ref('password'), null], 'Passwords must match'),
  }),
  handleSubmit: (payload, { setSubmitting, props }) => {
    // TODO: consider putting user in local storage
    const { user } = props;
    payload.roles = compact(payload.roles);
    api
      .put(`/users/${user.userId}`, payload, {
        withCredentials: true,
      })
      .then(res => {
        toast.success('Your user profile was updated successfully', {
          position: toast.POSITION.TOP_CENTER,
          hideProgressBar: true,
        });
      })
      .catch(err => {
        toast.error('Something went wrong', {
          position: toast.POSITION.TOP_CENTER,
          hideProgressBar: true,
        });
      });
    setSubmitting(false);
  },
  mapPropsToValues: ({ user }) => ({
    ...user,
  }),
  displayName: 'ProfileForm',
});

const Profile = props => {
  const modalProps = {
    triggerText: 'Forgot Password?',
  };

  const modalContent = <ForgotPassword user={{ email: '' }} {...props} />;

  const handleCheckAll = setFieldValue => {
    const inputs = document.getElementsByTagName('input');
    const selectAll = document.querySelector('#selectAll');
    const ids = { admin: 0, user: 1, family: 2, friend: 3 };
    for (const input of inputs) {
      if (Object.keys(ids).includes(input.id)) {
        if (selectAll.checked) {
          input.checked = true;
          setFieldValue(`roles[${ids[input.id]}]`, input.value);
        } else {
          input.checked = false;
          setFieldValue(`roles[${ids[input.id]}]`, '');
        }
      }
    }
  };

  const isAllChecked = () => {
    const inputs = document.getElementsByTagName('input');
    const ids = { admin: 0, user: 1, family: 2, friend: 3 };
    let isAllChecked = false;
    for (const input of inputs) {
      if (Object.keys(ids).includes(input.id)) {
        isAllChecked = input.checked;
        if (!isAllChecked) {
          break;
        }
      }
    }
    return isAllChecked;
  };

  document.title = 'User Profile';
  const {
    values,
    touched,
    errors,
    handleChange,
    handleBlur,
    handleSubmit,
    isSubmitting,
    isAdmin,
    setFieldValue,
  } = props;

  return (
    <React.Fragment>
      <h2>Profile</h2>
      <form onSubmit={handleSubmit}>
        <TextInput
          id="firstName"
          type="text"
          label="First Name"
          error={touched.firstName && errors.firstName}
          value={values.firstName}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        <TextInput
          id="lastName"
          type="text"
          label="Last Name"
          error={touched.lastName && errors.lastName}
          value={values.lastName}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        <TextInput
          id="email"
          type="email"
          label="Email"
          autoComplete="username email"
          error={touched.email && errors.email}
          value={values.email}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        <TextInput
          id="currentPassword"
          type="password"
          label="Current Password"
          autoComplete="current-password"
          error={touched.currentPassword && errors.currentPassword}
          value={values.currentPassword}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        <div className={styles.forgotPassword}>
          <Modal
            modalProps={modalProps}
            modalContent={modalContent}
            modalButtonClassName={commonStyles.modalLinkButton}
            {...props}
          />
        </div>
        <TextInput
          id="password"
          type="password"
          label="New Password"
          autoComplete="new-password"
          error={touched.password && errors.password}
          value={values.password}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        <TextInput
          id="confirmPassword"
          type="password"
          label="Confirm Password"
          autoComplete="new-password"
          error={touched.confirmPassword && errors.confirmPassword}
          value={values.confirmPassword}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        {isAdmin && (
          <React.Fragment>
            <h3>Roles</h3>
            {/* eslint-disable jsx-a11y/control-has-associated-label */}
            <input
              id="selectAll"
              type="checkbox"
              checked={isAllChecked()}
              onChange={() => handleCheckAll(setFieldValue)}
            />{' '}
            {/* eslint-enable jsx-a11y/control-has-associated-label */}
            <label htmlFor="selectAll">Select all</label>
            <CheckboxGroup
              id="roles"
              className={styles.roles}
              label="Which of these?"
              value={values.roles}
              onChange={setFieldValue}
            >
              <Field
                component={Checkbox}
                name="roles"
                id="admin"
                label="Admin"
                value="admin"
                defaultChecked={values.roles.includes('admin')}
              />
              <Field
                component={Checkbox}
                name="roles"
                id="user"
                label="User"
                value="user"
                defaultChecked={values.roles.includes('user')}
              />
              <Field
                component={Checkbox}
                name="roles"
                id="family"
                label="Family"
                value="family"
                defaultChecked={values.roles.includes('family')}
              />
              <Field
                component={Checkbox}
                name="roles"
                id="friend"
                label="Friend"
                value="friend"
                defaultChecked={values.roles.includes('friend')}
              />
            </CheckboxGroup>
          </React.Fragment>
        )}
        <button type="submit" disabled={isSubmitting} className="btn btn-primary">
          Submit
        </button>
      </form>
    </React.Fragment>
  );
};

export default withAuthentication(formikEnhancer(Profile));
