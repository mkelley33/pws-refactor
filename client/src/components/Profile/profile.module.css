.roles input {
  margin-right: .33rem;
}

.forgotPassword {
  font-size: .8em;
  margin-top: -.66rem;
  margin-bottom: 1.33em;
  margin-left: .66em;
}
