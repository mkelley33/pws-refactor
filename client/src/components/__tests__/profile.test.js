import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, waitFor } from '@testing-library/react';
import mockAxios from 'axios';
import Profile from '../Profile/Profile';

describe('Profile', () => {
  // Pass a fake user with formik default values for use in mapPropsToValues
  const user = { userId: 1, firstName: '', lastName: '', email: '', password: '', confirmPassword: '' };

  beforeEach(() => {
    mockAxios.doMockReset();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Profile />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('submits form successfully', async () => {
    const { getByText, getByLabelText } = render(<Profile user={user} />);

    const firstName = getByLabelText(/first name/i);
    const lastName = getByLabelText(/last name/i);
    const email = getByLabelText(/email/i);
    const submit = getByText('Submit');

    await waitFor(() => {
      fireEvent.change(firstName, {
        target: {
          value: 'Michaux',
        },
      });
    });

    await waitFor(() => {
      fireEvent.change(lastName, {
        target: {
          value: 'Kelley',
        },
      });
    });

    await waitFor(() => {
      fireEvent.change(email, {
        target: {
          value: 'test@test.com',
        },
      });
    });

    fireEvent.click(submit);

    await waitFor(() => {
      expect(mockAxios.put).toHaveBeenCalledTimes(1);
    });
  });

  it('validates the first name cannot be blank', async () => {
    const { getByLabelText, getByText } = render(<Profile />);
    const firstName = getByLabelText(/first name/i);

    fireEvent.blur(firstName);

    let error;
    await waitFor(() => {
      error = getByText('First name is required');
    });

    expect(error).not.toBeNull();
  });

  it('validates the last name cannot be blank', async () => {
    const { getByLabelText, getByText } = render(<Profile />);
    const lastName = getByLabelText(/last name/i);

    fireEvent.blur(lastName);

    let error;
    await waitFor(() => {
      error = getByText('Last name is required');
    });

    expect(error).not.toBeNull();
  });

  it('validates email cannot be blank', async () => {
    const { getByLabelText, getByText } = render(<Profile />);
    const email = getByLabelText(/email/i);

    fireEvent.blur(email);

    let error;
    await waitFor(() => {
      error = getByText('Email is required');
    });

    expect(error).not.toBeNull();
  });

  it('validates that first name must be longer than one character', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const firstName = getByLabelText(/first name/i);

    await waitFor(() => {
      fireEvent.change(firstName, {
        target: {
          value: 'M',
        },
      });
    });

    fireEvent.blur(firstName);

    let error;
    await waitFor(() => {
      error = getByText("C'mon, your first name is longer than that");
    });

    expect(error).not.toBeNull();
  });

  it('validates that last name must be longer than one character', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const lastName = getByLabelText(/last name/i);

    await waitFor(() => {
      fireEvent.change(lastName, {
        target: {
          value: 'K',
        },
      });
    });

    fireEvent.blur(lastName);

    let error;
    await waitFor(() => {
      error = getByText("C'mon, your last name is longer than that");
    });

    expect(error).not.toBeNull();
  });

  it('validates email address must be correctly formatted', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const email = getByLabelText(/email/i);

    await waitFor(() => {
      fireEvent.change(email, {
        target: {
          value: 'email at dot com',
        },
      });
    });

    fireEvent.blur(email);

    let error;
    await waitFor(() => {
      error = getByText('Invalid email address');
    });

    expect(error).not.toBeNull();
  });

  it('validates that new password must be at least 8 characters long', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'p',
        },
      });
    });

    fireEvent.blur(password);

    let error;
    await waitFor(() => {
      error = getByText('Password has to be at least 8 characters!');
    });

    expect(error).not.toBeNull();
  });

  it('validates that new password must contain at least one uppercase letter', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'asdfjkl;3',
        },
      });
    });

    fireEvent.blur(password);

    let error;
    await waitFor(() => {
      error = getByText(
        'Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character'
      );
    });

    expect(error).not.toBeNull();
  });

  it('validates that new password must contain at least one lowercase letter', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'ASDFJKL;3',
        },
      });
    });

    fireEvent.blur(password);

    let error;
    await waitFor(() => {
      error = getByText(
        'Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character'
      );
    });

    expect(error).not.toBeNull();
  });

  it('validates that new password must contain at least one number', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'ASDFJKLa;',
        },
      });
    });

    fireEvent.blur(password);

    let error;
    await waitFor(() => {
      error = getByText(
        'Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character'
      );
    });

    expect(error).not.toBeNull();
  });

  it('validates that new password must contain at least one special character', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'ASDFJKLa3',
        },
      });
    });

    fireEvent.blur(password);

    let error;
    await waitFor(() => {
      error = getByText(
        'Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character'
      );
    });

    expect(error).not.toBeNull();
  });

  it('validates a well-formatted password', async () => {
    const { getByLabelText, queryByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'ASDFJKLa3!',
        },
      });
    });

    fireEvent.blur(password);

    let error;
    await waitFor(() => {
      error = queryByText(
        'Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character'
      );
    });

    expect(error).toBeNull();
  });

  it('validates that confirm password matches new password', async () => {
    const { getByLabelText, getByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);
    const confirmPassword = getByLabelText(/confirm password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'ASDFJKLa3!',
        },
      });
      fireEvent.change(confirmPassword, {
        target: {
          value: 'ASDFJKLa3',
        },
      });
    });

    fireEvent.blur(confirmPassword);

    let error;
    await waitFor(() => {
      error = getByText('Passwords must match');
    });

    expect(error).not.toBeNull();
  });

  it('validates succesfully that confirm password matches new password', async () => {
    const { getByLabelText, queryByText } = render(<Profile user={user} />);
    const password = getByLabelText(/new password/i);
    const confirmPassword = getByLabelText(/confirm password/i);

    await waitFor(() => {
      fireEvent.change(password, {
        target: {
          value: 'ASDFJKLa3!',
        },
      });
      fireEvent.change(confirmPassword, {
        target: {
          value: 'ASDFJKLa3',
        },
      });
    });

    fireEvent.blur(confirmPassword);

    let error;
    await waitFor(() => {
      error = queryByText('Passwords must match');
    });

    expect(error).toBeNull();
  });
});
