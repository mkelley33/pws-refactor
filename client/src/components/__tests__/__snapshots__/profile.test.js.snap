// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
Array [
  <h2>
    Profile
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className="sr-only"
        htmlFor="firstName"
      >
        First Name
      </label>
      <input
        className="form-control"
        id="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="First Name"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="sr-only"
        htmlFor="lastName"
      >
        Last Name
      </label>
      <input
        className="form-control"
        id="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Last Name"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="sr-only"
        htmlFor="email"
      >
        Email
      </label>
      <input
        autoComplete="username email"
        className="form-control"
        id="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        type="email"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="sr-only"
        htmlFor="currentPassword"
      >
        Current Password
      </label>
      <input
        autoComplete="current-password"
        className="form-control"
        id="currentPassword"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Current Password"
        type="password"
      />
    </div>
    <div
      className="forgotPassword"
    >
      <button
        className="modalLinkButton"
        onClick={[Function]}
        type="button"
      >
        Forgot Password?
      </button>
    </div>
    <div
      className="form-group"
    >
      <label
        className="sr-only"
        htmlFor="password"
      >
        New Password
      </label>
      <input
        autoComplete="new-password"
        className="form-control"
        id="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="New Password"
        type="password"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="sr-only"
        htmlFor="confirmPassword"
      >
        Confirm Password
      </label>
      <input
        autoComplete="new-password"
        className="form-control"
        id="confirmPassword"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Confirm Password"
        type="password"
      />
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>,
]
`;
